diff --git a/misc.c b/misc.c
index d40f9901..454499e8 100644
--- a/misc.c
+++ b/misc.c
@@ -286,7 +286,8 @@ char *dupstr(const char *s)
 {
     char *p = NULL;
     if (s) {
-        int len = strlen(s);
+        size_t len = strlen(s);
+        assert(len < SIZE_MAX);
         p = snewn(len + 1, char);
         strcpy(p, s);
     }
@@ -456,17 +457,19 @@ char *dupvprintf(const char *fmt, va_list ap)
 char *fgetline(FILE *fp)
 {
     char *ret = snewn(512, char);
-    int size = 512, len = 0;
+    size_t size = 512, len = 0;
     while (fgets(ret + len, size - len, fp)) {
 	len += strlen(ret + len);
 	if (len > 0 && ret[len-1] == '\n')
 	    break;		       /* got a newline, we're done */
 	size = len + 512;
+        assert(size < INT_MAX);
 	ret = sresize(ret, size, char);
+        assert(ret);
     }
-    if (len == 0) {		       /* first fgets returned NULL */
-	sfree(ret);
-	return NULL;
+    if (len == 0) { /* first fgets returned NULL */
+        sfree(ret);
+        return NULL;
     }
     ret[len] = '\0';
     return ret;
@@ -742,6 +745,7 @@ void *safemalloc(size_t n, size_t size)
     if (fp)
 	fprintf(fp, "malloc(%d) returns %p\n", size, p);
 #endif
+    assert(p);
     return p;
 }
 
@@ -753,6 +757,8 @@ void *saferealloc(void *ptr, size_t n, size_t size)
 	p = NULL;
     } else {
 	size *= n;
+        if (!size)
+            size = 1;
 	if (!ptr) {
 #ifdef MINEFIELD
 	    p = minefield_c_malloc(size);
@@ -784,6 +790,7 @@ void *saferealloc(void *ptr, size_t n, size_t size)
     if (fp)
 	fprintf(fp, "realloc(%p,%d) returns %p\n", ptr, size, p);
 #endif
+    assert(p);
     return p;
 }
 
@@ -871,6 +878,8 @@ int conf_launchable(Conf *conf)
 {
     if (conf_get_int(conf, CONF_protocol) == PROT_SERIAL)
 	return conf_get_str(conf, CONF_serline)[0] != 0;
+    else if (conf_get_int(conf, CONF_protocol) == PROT_CYGTERM)
+        return conf_get_str(conf, CONF_cygcmd)[0] != 0;
     else
 	return conf_get_str(conf, CONF_host)[0] != 0;
 }
@@ -879,6 +888,8 @@ char const *conf_dest(Conf *conf)
 {
     if (conf_get_int(conf, CONF_protocol) == PROT_SERIAL)
 	return conf_get_str(conf, CONF_serline);
+    else if (conf_get_int(conf, CONF_protocol) == PROT_CYGTERM)
+        return conf_get_str(conf, CONF_cygcmd);
     else
 	return conf_get_str(conf, CONF_host);
 }
