diff --git a/pscp.c b/pscp.c
index dc9e1f50..03e57218 100644
--- a/pscp.c
+++ b/pscp.c
@@ -382,7 +382,7 @@ static void do_cmd(char *host, char *user, char *cmd)
 	/* Try to load settings for `host' into a temporary config */
 	Conf *conf2 = conf_new();
 	conf_set_str(conf2, CONF_host, "");
-	do_defaults(host, conf2);
+	do_defaults_then_file(host, conf2);
 	if (conf_get_str(conf2, CONF_host)[0] != '\0') {
 	    /* Settings present and include hostname */
 	    /* Re-load data into the real config. */
