diff --git a/sshcrcda.c b/sshcrcda.c
index 8d77cbb6..5518accf 100644
--- a/sshcrcda.c
+++ b/sshcrcda.c
@@ -153,6 +153,7 @@ int detect_attack(void *handle, uchar *buf, uint32 len, uchar *IV)
         for (i = HASH(c) & (ctx->n - 1); ctx->h[i] != HASH_UNUSED;
              i = (i + 1) & (ctx->n - 1)) {
             if (ctx->h[i] == HASH_IV) {
+                assert(IV);
                 if (!CMP(c, IV)) {
                     if (check_crc(c, buf, len, IV))
                         return 1;      /* attack detected */
