diff --git a/sshpubk.c b/sshpubk.c
index a4ecb9d5..e547c3c8 100644
--- a/sshpubk.c
+++ b/sshpubk.c
@@ -141,7 +141,6 @@ static int loadrsakey_main(FILE * fp, struct RSAKey *key, int pub_only,
     i += j;
     j = ssh1_read_bignum(buf + i, len - i, &key->p);
     if (j < 0) goto end;
-    i += j;
 
     if (!rsa_verify(key)) {
 	*error = "rsa_verify failed";
@@ -1181,12 +1180,12 @@ int key_type(const Filename *filename)
 char *key_type_to_str(int type)
 {
     switch (type) {
-      case SSH_KEYTYPE_UNOPENABLE: return "unable to open file"; break;
-      case SSH_KEYTYPE_UNKNOWN: return "not a private key"; break;
-      case SSH_KEYTYPE_SSH1: return "SSH-1 private key"; break;
-      case SSH_KEYTYPE_SSH2: return "PuTTY SSH-2 private key"; break;
-      case SSH_KEYTYPE_OPENSSH: return "OpenSSH SSH-2 private key"; break;
-      case SSH_KEYTYPE_SSHCOM: return "ssh.com SSH-2 private key"; break;
-      default: return "INTERNAL ERROR"; break;
+      case SSH_KEYTYPE_UNOPENABLE: return "unable to open file";
+      case SSH_KEYTYPE_UNKNOWN: return "not a private key";
+      case SSH_KEYTYPE_SSH1: return "SSH-1 private key";
+      case SSH_KEYTYPE_SSH2: return "PuTTY SSH-2 private key";
+      case SSH_KEYTYPE_OPENSSH: return "OpenSSH SSH-2 private key";
+      case SSH_KEYTYPE_SSHCOM: return "ssh.com SSH-2 private key";
+      default: return "INTERNAL ERROR";
     }
 }
