diff --git a/unix/uxmisc.c b/unix/uxmisc.c
index b7727cb2..2a70937d 100644
--- a/unix/uxmisc.c
+++ b/unix/uxmisc.c
@@ -290,3 +290,7 @@ FontSpec *fontspec_deserialise(void *vdata, int maxsize, int *used)
     *used = end - data + 1;
     return fontspec_new(data);
 }
+
+void sanitise_path_leaving_slashes(Filename *original, char *path) {
+    // nothing to do here; everything is typically legal
+}
