diff --git a/windows/winutils.c b/windows/winutils.c
index ef0db921..a0c5dd16 100644
--- a/windows/winutils.c
+++ b/windows/winutils.c
@@ -147,7 +147,10 @@ void pgp_fingerprints(void)
                 "Original PuTTY Master Key (RSA, 1024-bit):\n"
                 "  " PGP_RSA_MASTER_KEY_FP "\n"
                 "Original PuTTY Master Key (DSA, 1024-bit):\n"
-                "  " PGP_DSA_MASTER_KEY_FP,
+                "  " PGP_DSA_MASTER_KEY_FP "\n"
+                "PuTTYTray is signed by the unrelated:\n"
+                "Chris West (Faux) <gpg@goeswhere.com>:\n"
+                "  408A E4F1 4EA7 33EF 1265  82C1 B195 E1C4 779B A9B2\n",
 		"PGP fingerprints", MB_ICONINFORMATION | MB_OK,
 		HELPCTXID(pgp_fingerprints));
 }
