--- /dev/null
+++ b/windows/putty64.mft
@@ -0,0 +1,31 @@
+<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
+<!-- Do not attempt to do anything clever with this file, as some versions of
+     Windows are very sensitive to the exact format.
+     Hence, some facts below are fibs. -->
+<assembly xmlns="urn:schemas-microsoft-com:asm.v1" manifestVersion="1.0">
+<assemblyIdentity
+   version="0.0.0.0"
+   processorArchitecture="amd64"
+   name="PuTTY"
+   type="win32" />
+   <description>A network client and terminal emulator</description>
+   <dependency>
+   <dependentAssembly>
+        <!-- Load Common Controls 6 instead of 5 to get WinXP native-
+             looking controls in the client area. -->
+        <assemblyIdentity type="win32"
+             name="Microsoft.Windows.Common-Controls"
+             version="6.0.0.0" 
+             publicKeyToken="6595b64144ccf1df"
+             language="*" 
+             processorArchitecture="amd64"/> 
+   </dependentAssembly>
+   </dependency>
+   <!-- Declare us to be "DPI-aware". -->
+   <asmv3:application xmlns:asmv3="urn:schemas-microsoft-com:asm.v3">
+     <asmv3:windowsSettings
+         xmlns="http://schemas.microsoft.com/SMI/2005/WindowsSettings">
+       <dpiAware>true</dpiAware>
+     </asmv3:windowsSettings>
+   </asmv3:application>
+</assembly>
diff --git a/windows/puttycfg.ico b/windows/puttycfg.ico
index bb32be84..f53d56a8 100644
Binary files a/windows/puttycfg.ico and b/windows/puttycfg.ico differ
diff --git a/windows/puttycyg.ico b/windows/puttycyg.ico
new file mode 100644
index 00000000..ea62d103
Binary files /dev/null and b/windows/puttycyg.ico differ
