--- /dev/null
+++ b/windows/cthelper/TODO
@@ -0,0 +1,38 @@
+* cthelper
+
+[done] A way to propagate fatal error messages back to PuTTYcyg to display in a message box
+
+BUG: Duplicate session brings up dialog on cygterm terminals using the user's default shell.
+
+BUG: When running an arbitrary command, PuTTYcyg closes one character after EOF.
+
+Misfeature: In Cygwin ps output, the cthelper process does not show with which TTY it is associated.
+
+Flow control stop (^S) should be more instantaneous
+
+* PuTTY cygterm backend
+
+hide the Port editbox
+
+[done] allow setting a default command of "-" or allow empty command to launch the user's shell
+
+* PuTTY (not related to cygterm backend)
+
+[done] add Ctrl-/ map to C-_
+
+[done] add Shift-Enter, Ctrl-Enter map to LINEFEED
+
+[done] add Ctrl-Backspace map same as Shift-Backspace (alternate 0x08 or 0x7F)
+
+[done] PuTTY wish alt-metabit
+
+add Alt modifier to mouse location reporting (low-priority)
+
+add ssh-agent support to PuTTY (?) -or- add pageant support to openssh (Cygwin)
+
+add cursor styles vertical-half-block and horizontal-half-block 
+
+PuTTY wish key-mapping
+
+PuTTY wish portable-keyboard
+
diff --git a/windows/cthelper/buffer.c b/windows/cthelper/buffer.c
new file mode 100644
index 00000000..01be6c66
